magtogoek/tools.py

```diff
@@ -180,4 +181,307 @@ def rotate_2d_vector(
+    elements :
+        Elements to look for (k by 1).
+    axis :
+        Index of the dimension along which look for `elements` in `array`.
+
+    Returns
+    -------
+    bool_array :
+        True if element was found, False otherwise. Of same size as
+        `array` except dimension `axis`. For example, if `axis = 1`
+        `bool_array` is of size m by k by l by (...). Dimension `axis`
+        `bool_array` represents if element was found in `array` along
+        dimension `axis` for each element in `elements`.
+
+    """
+    # Compute result for ndarray
+    _func = lambda a, elements: np.isin(elements, a)
+    result = np.apply_along_axis(_func, axis, array, elements)
+
+    # Form into DataArray if input is DataArray
+    if isinstance(array, xr.DataArray):
+        shape, dims = [], []
+        for i_, (dim_size, dim_name) in enumerate(zip(array.shape, array.dims)):
+            if i_ != axis:
+                shape.append(dim_size)
+                dims.append(dim_name)
+        result = xr.DataArray(result.reshape(*shape), dims=dims)
+
+    return result
+
+
+def _new_flags_bin_regrid(flags: xr.DataArray,
+                          bin_depths: np.ndarray,
+                          dim: str = 'depth',
+                          axis: int = 0) -> xr.DataArray:
+    """
+    Quality flag transfer function for bin average regridding.
+
+    Parameters
+    ----------
+    flags :
+        Flags prepared for regridding (5, 8, 9).
+    bin_depths :
+        Center of the regridding averaging bins (meters).
+    dim :
+        Name of the dimension along which the binning is done.
+    axis :
+        Index of the dimension along which the binning is done.
+
+    Returns
+    -------
+    new_flags :
+        Quality flags for the regridded data.
+
+    """
+    bin_edges = _bin_centers_to_edges(bin_depths)
+
+    # Determine if bins contained 5 or 8 flags
+    grouped = flags.groupby_bins(dim, bin_edges, labels=bin_depths)
+    contains_8 = grouped.map(_isin_along_axis, args=(8, axis,))
+    contains_5 = grouped.map(_isin_along_axis, args=(5, axis,))
+
+    # Make new flag array
+    new_flags = xr.where(contains_5, 5, 9)  # contains 5 or 8 -> 5, not -> 9
+    new_flags = xr.where(contains_8, 8, new_flags)  # contains 8 -> 8
+
+    return new_flags
+
+
+def _new_flags_interp_regrid(dataset: xr.Dataset, variable: str) -> xr.DataArray:
+    """
+    Quality flag transfer function for interpolation regridding.
+
+    Parameters
+    ----------
+    flags :
+        Dataset being regridded.
+    variable :
+        Name of the variable for which flags are being transfered.
+
+    Returns
+    -------
+    new_flags :
+        Quality flags for the regridded data variable.
+
+    """
+    condition_null = dataset[f"{variable}"].isnull()
+    condition_good = (dataset[f"{variable}_QC"] == 8) & ~condition_null
+    condition_changed = (dataset[f"{variable}_QC"] == 5) & condition_null
+    new_flags = dataset[f"{variable}_QC"].copy()
```

Comment:
If nans are propagating during the interpolation, we don't need to check the interp flags to finds good values since with change all values to nans except where the flag in [0,1,2].
```suggestion
    good_condition = dataset[f"{variable}"].isfinite()
    condition_changed = dataset[f"{variable}_QC"] == 5
    new_flags = dataset[f"{variable}_QC"].copy()
```